src/qat/purr/compiler/execution.py

```diff
@@ -453,6 +453,14 @@ def create_duration_timeline(self, instructions: List[QuantumInstruction]):
         total_durations: Dict[PulseChannel, int] = dict()
 
         for instruction in instructions:
+            # calculate the total durations for each component
+            if isinstance(instruction, Synchronize):
+                current_durations = [
+                    total_durations.setdefault(qt, 0) for qt in instruction.quantum_targets
+                ]
+                longest_length = max(current_durations, default=0.0)
+                ctr = 0  # count to eliminate dict search later
```

Comment:
how do you explain the speedup by just using `ctr` as index ?

Reply:
> it might be redundant later if we can cache the full id?
correct. keys in the original dictionary (`total_durations`) are pulse channels. When you read a key, the pulse channel gets hashed which gets `full_id` to be called. So what you're doing here is circumvent hashing pulse channels to hashing just integers, which becomes trivial.

The overall improvement here shows one beneficial side of caching `full_id`.

Reply:
Yeah good point! I guess for readability, it would be better to keep `current_durations` as a dict with the hope that these issues are addressed later with caching.

Note that it is still worth calculating `longest_length` outside of the loop (in practice, even with the dict and caching, I see some speed-up by doing this).